1. Найти с помощью NumPy SVD для матрицы

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$


In [1]:
import numpy as np
np.set_printoptions(precision=5, suppress=True)

In [2]:
A = np.array([[1, 2, 0],
              [0, 0, 5],
              [3,-4, 2],
              [1, 6, 5],
              [0, 1, 0]])
print(f'Матрица A:\n{A}')

Матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [3]:
U, s, W = np.linalg.svd(A)

# Транспонируем матрицу W
V = W.T

# s - список диагональных элементов, его нужно привести к виду диагональной матрицы для наглядности
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

In [4]:
print(f'Матрица D:\n{D}')

Матрица D:
[[8.82487 0.      0.     ]
 [0.      6.14061 0.     ]
 [0.      0.      2.53272]
 [0.      0.      0.     ]
 [0.      0.      0.     ]]


In [5]:
print(f'Матрица U:\n{U}')

Матрица U:
[[ 0.17057  0.15681 -0.53078 -0.79905 -0.16158]
 [ 0.39287 -0.52934  0.61348 -0.43376  0.03082]
 [-0.14366 -0.82449 -0.52379  0.1405   0.074  ]
 [ 0.88844  0.06074 -0.24655  0.37756 -0.06043]
 [ 0.08125  0.10832 -0.08231 -0.10525  0.98174]]


In [6]:
# Убедимся, что она действительно ортогональна
print(np.dot(U.T, U))

[[ 1.  0. -0. -0. -0.]
 [ 0.  1.  0. -0.  0.]
 [-0.  0.  1. -0. -0.]
 [-0. -0. -0.  1. -0.]
 [-0.  0. -0. -0.  1.]]


In [7]:
print(f'Матрица V:\n{V}')

Матрица V:
[[ 0.07116 -0.36738 -0.92735]
 [ 0.71702  0.66514 -0.20848]
 [ 0.69341 -0.65009  0.31075]]


In [8]:
# Убедимся, что она действительно ортогональна
print(np.dot(V.T, V))

[[ 1.  0.  0.]
 [ 0.  1. -0.]
 [ 0. -0.  1.]]


In [9]:
# Проведем проверку
print(np.dot(np.dot(U, D), V.T))

[[ 1.  2.  0.]
 [ 0. -0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [-0.  1.  0.]]




2. Для матрицы из предыдущего задания найти:

    а) евклидову норму;
    
    б) норму Фробениуса.

а) $\left \| A \right \|_{E}=\mu_{1} = 8.82487.$


б) $$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}}.$$

In [10]:
A_f =  np.sqrt(sum(item**2 for item in s))
A_f 

11.045361017187261

$\left \| A \right \|_{F}= 11.045361017187261.$